<a href="https://colab.research.google.com/github/HannahParker/cse380-notebooks/blob/master/09_2_Ponder_and_Prove_Data_Compression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Data Compression
#### Due: Saturday, 6 March 2021, 11:59 pm.

# TODO Explore Huffman Trees and Huffman Codes


Your task is examine how to compress a *special piece of information* as compactly as possible, and **calculate various compression ratios**.

Recall that the **compression ratio** of a variable-length encoding like Huffman encoding is the percentage $100(f - v)/f$, where $f$ is the number of bits per symbol of the smallest **fixed**-length encoding, and $v$ is the average number of bits per symbol with the variable-length encoding.

For example, if there were 9 different symbols in a message, $f=4$ is the number of bits of the smallest fixed-length encoding, because $2^3 = 8$ (not enough for $9$) and $2^4 = 16$ (enough and to spare). If the variable-length encoding of the message had $v=3.12$, the compression ratio would be $100(4 - 3.12)/4 \approx 22\%$.

Note that calculating the average number of bits per symbol is not strictly necessary. That's because an alternate and equivalent way is to calculate $100(ft - vt)/ft$, where $ft$ is the **total** number of bits encoded with the fixed encoding, and $vt$ is the **total** number of bits encoded with the variable-length encoding.

The *special piece of information* to be compressed is a list of the first ten million primes. This is a list that starts

|    |
|----|
|  2 |
|  3 |
|  5 |
|  7 |
| 11 |
| 13 |
| 17 |
| 19 |
| 23 |
| 29 |

  and ends

|           |
|-----------|
| 179424551 |
| 179424571 |
| 179424577 |
| 179424601 |
| 179424611 |
| 179424617 |
| 179424629 |
| 179424667 |
| 179424671 |
| 179424673 |

As ASCII text stored in a file with one prime per line, the size of this data file is slightly over 89 megabytes. The goal is to compress this down to just over 5 megabytes (5589056 bytes, to be exact). That's a 94% compression ratio!

Standard compression tools can only get about a 73% compression ratio for this ASCII data. A more clever approach is needed. Instead of compressing the list of prime numbers, compress a list of the *gaps* between them!

It doesn't save much, just the unique (occurring only once) gap size of 1 between 2 and 3, but in the spirit of de Polignac's conjecture that every *even* number appears infinitely often as a gap between consecutive primes, just consider the even-sized gaps. The result will be a list that starts with 2 (the difference between 5 and 3), 2 (the difference between 7 and 5), 4 (the difference between 11 and 7), 2 (the difference between 13 and 11), 4 (the difference between 17 and 13), 2 (the difference between 19 and 17), 4 (the difference between 23 and 19), and 6 (the difference between 29 and 23).

Generating this data is the first task. The algorithm for doing so is very straightforward:

1. Find the gaps between consecutive odd primes.
2. Store these gaps as a list of even numbers.

Tabulating the results, the first ten gaps and the last ten gaps are as follows, where the numbers after the equals signs are the gaps to list:

|                 |
|-----------------|
|  5  -   3  =  2 |
|  7  -   5  =  2 |
| 11  -   7  =  4 |
| 13  -  11  =  2 |
| 17  -  13  =  4 |
| 19  -  17  =  2 |
| 23  -  19  =  4 |
| 29  -  23  =  6 |
| 31  -  29  =  2 |
| 37  -  31  =  6 |

|                                |
|--------------------------------|
| 179424551  -  179424533  =  18 |
| 179424571  -  179424551  =  20 |
| 179424577  -  179424571  =   6 |
| 179424601  -  179424577  =  24 |
| 179424611  -  179424601  =  10 |
| 179424617  -  179424611  =   6 |
| 179424629  -  179424617  =  12 |
| 179424667  -  179424629  =  38 |
| 179424671  -  179424667  =   4 |
| 179424673  -  179424671  =   2 |

As a correctness check, see if your generated list of gaps has length 9999998.

The next step is to count how many times each gap size occurs, so that for the Huffman encoding scheme, the larger the frequency of occurrence, the smaller the number of bits encoding that gap size.

As a correctness check, here are the first ten and the last ten gap counts:

|  Gap | Count   |
|------|---------|
|    2 |  738597 |
|    4 |  738717 |
|    6 | 1297540 |
|    8 |  566151 |
|   10 |  729808 |
|   12 |  920661 |
|   14 |  503524 |
|   16 |  371677 |
|   18 |  667734 |
|   20 |  354267 |
|      |         |
|  190 |       1 |
|  192 |       3 |
|  194 |       1 |
|  196 |       1 |
|  198 |       6 |
|  202 |       2 |
|  204 |       3 |
|  210 |       4 |
|  220 |       1 |
|  222 |       1 |

Note two things from these partial gap counts:

1. Small even numbers (< 100) are well represented, larger ones (< 1000) less so.
2. Ten million primes aren't enough to have *every* even number represented; for example, 200, 206, 208, 212, 214, 216, and 218 do not appear even once.


In [6]:
from sympy import primerange

ten_millionth_prime = 179424673

first_ten_million_primes = list(primerange(1, ten_millionth_prime + 1))

print(first_ten_million_primes[0:200])

def list_the_gaps(n=first_ten_million_primes):
    return list(map(lambda i: n[i]-n[i-1],range(2,len(n))))

def restore_primes(list_of_gaps):
    prime = 3
    plist = [2, 3]
    for gap in list_of_gaps:
        prime += gap
        plist.append(prime)
    return plist

list_of_gaps = list_the_gaps()
length_of_list_of_gaps = len(list_of_gaps)
print("Length of list of gaps:")
print(length_of_list_of_gaps)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 12

In [ ]:
first_ten_million_primes_restored = restore_primes(list_of_gaps)
first_ten_million_primes_restored == first_ten_million_primes

True

In [ ]:
from sympy import primerange

table = {}

prev = 3
gap = 0
for prime in list(primerange(4, ten_millionth_prime + 1)):
    gap = prime - prev
    #print(gap)
    prev = prime
    if (gap in table):
        table[gap] += 1
    else:
        table[gap] = 1

table

{2: 738597,
 4: 738717,
 6: 1297540,
 8: 566151,
 10: 729808,
 12: 920661,
 14: 503524,
 16: 371677,
 18: 667734,
 20: 354267,
 22: 307230,
 24: 453215,
 26: 211203,
 28: 229177,
 30: 398713,
 32: 123123,
 34: 129043,
 36: 206722,
 38: 94682,
 40: 111546,
 42: 159956,
 44: 64866,
 46: 54931,
 48: 93693,
 50: 52183,
 52: 38800,
 54: 64157,
 56: 32224,
 58: 27985,
 60: 55305,
 62: 16763,
 64: 17374,
 66: 30960,
 68: 12368,
 70: 17475,
 72: 17255,
 74: 8540,
 76: 7253,
 78: 13758,
 80: 6760,
 82: 4791,
 84: 9818,
 86: 3411,
 88: 3454,
 90: 7056,
 92: 2259,
 94: 2058,
 96: 3544,
 98: 1831,
 100: 1923,
 102: 2374,
 104: 1168,
 106: 933,
 108: 1634,
 110: 941,
 112: 711,
 114: 1125,
 116: 439,
 118: 433,
 120: 948,
 122: 287,
 124: 318,
 126: 533,
 128: 183,
 130: 211,
 132: 301,
 134: 128,
 136: 100,
 138: 210,
 140: 140,
 142: 90,
 144: 123,
 146: 46,
 148: 67,
 150: 94,
 152: 52,
 154: 43,
 156: 57,
 158: 19,
 160: 27,
 162: 27,
 164: 20,
 166: 9,
 168: 25,
 170: 18,
 172: 4,
 174: 10,
 1

# TODO Determine Exact Size of Data to be Compressed


Without actually doing it, imagine creating an ASCII file containing the first ten million primes, represented in decimal, one prime per line. Calculate the size of this file, so you can show an exceptional compression ratio from it (see below).

Using a binary encoding instead of ASCII, each prime requires 32 bits (4 bytes), so the size of a binary file is easily determined.

Using a fixed-width encoding of the gap counts, however, requires knowing how many different gap sizes there are, after which the calculation is straightforward.

In [2]:
!pip install pyprimesieve

  Created wheel for pyprimesieve: filename=pyprimesieve-0.1.6-cp37-cp37m-linux_x86_64.whl size=368927 sha256=26674ae5becc645580ec27fc4d41ce6972b6e9f5615af382b9bdd16c7d95c529
  Stored in directory: /root/.cache/pip/wheels/c4/63/2b/a485079de882a375d28a4dc141386c76ea9a6aaad505f2198b
Successfully built pyprimesieve


In [3]:
import pyprimesieve

# Thanks, Kyle!
tmp = pyprimesieve.primes_nth(10000000)
primes = pyprimesieve.primes(tmp+1)
gaps = [*map(lambda i:primes[i]-primes[i-1],range(2,10000000))]
from math import log10, floor

# get the number of digits in a certain number
def get_num_digits_in_num(n):
    return floor(log10(n))

# make sure to calculate in the newline character
def get_line_size(n):
    return get_num_digits_in_num(n) + 1
total_size_in_digits = sum(map(lambda p: get_line_size(p), primes))
total_size_in_bits = total_size_in_digits * 8
print("File size in Megabytes:", (total_size_in_digits + 10 ** 7) / 2 ** 20)


File size in Megabytes: 89.15371894836426


In [16]:
from math import floor, log2
floor(log2(max(gaps))) + 1
from math import ceil
ceil(log2(len(set(gaps))))

7

In [20]:
a_set = set(list_of_gaps)
number_of_unique_values = len(a_set)
print("Unique Gap Sizes:", number_of_unique_values)

Unique Gap Sizes: 104


# TODO Use Functional Python


You are encouraged to use the [anytree](https://pypi.org/project/anytree) Python library, which has a nice exporter by way of which you can graphically view trees. (You may recall using this in DM1, and thus know that **anytree** depends on [graphviz](https://graphviz.org), which you also used.)

This library uses the object-oriented features of Python to create and visualize trees. You are encouraged to use the functional features of Python as much as possible, achieving your results not by using some existing third-party libraries for building Huffman Trees and Codes, but writing your own code as cleanly and elegantly as you can.

In [11]:
!pip install anytree
from anytree import Node, RenderTree, PreOrderIter, Walker
from anytree.util import leftsibling, rightsibling
from anytree.exporter.dotexporter import DotExporter
from collections import Counter
from queue import PriorityQueue
from math import ceil, floor, log
from sympy import primerange

def realprimes_up_to(n):
  return list(primerange(4, n))

def get_list_of_gaps(pl):
  gaps_list = list(map(lambda i: pl[i] - pl[i - 1], range(1, len(pl))))
  gaps_list = [2] + gaps_list # [2] for the gap between 3 and 5
  return gaps_list

class GapNode(Node):
  def __lt__(self, other):
    return self.count < other.count

node_counter = 0
def next_node_name():
  global node_counter
  name = 'gn' + str(node_counter)
  node_counter += 1
  return name

def new_node(gp, ct):
  return GapNode(next_node_name(), gap = gp, count = ct)

def new_internal_node(left, right):
  return GapNode(next_node_name(), children = [left, right],
                 gap = 0, count = left.count + right.count)

def make_huffman_tree(gaps_list):
  gap_dict = Counter(gaps_list)
  q = PriorityQueue()
  for key, val in gap_dict.items():
    q.put(new_node(key, val))

  while q.qsize() > 1:
    left = q.get()
    right = q.get()
    q.put(new_internal_node(left, right))

  return q.get()

def get_codes(root):
  leaves = [node for node in PreOrderIter(root, filter_=lambda n: not n.children)]
  codes = {}
  w = Walker()
  for leaf in leaves:
    path = w.walk(leaf, root)[0]
    code = ''
    for node in path:
      code = ('1' if leftsibling(node) else '0') + code
    codes[leaf.gap] = tuple([code, leaf.count])
  return codes

def compression_ratio(f, v):
  return 100 * (f - v) / f

def get_encoded_size(codes):
  return sum([len(code) * count for gap, (code, count) in codes.items()])

def get_fixed_size(codes):
  num_keys = len(codes)
  num_bits_per_key = ceil(log(num_keys, 2))
  return sum([num_bits_per_key * count for gap, (code, count) in codes.items()])

def report(codes):
  return compression_ratio(get_fixed_size(codes), get_encoded_size(codes))

def test_up_to(primes, upper, results):
  list_of_gaps = get_list_of_gaps(primes[:upper])
  print(primes[:upper])
  print(list_of_gaps)
  root = make_huffman_tree(list_of_gaps)
  print(RenderTree(root))
  DotExporter(root).to_picture(f'gap-tree-{upper:02d}.png')
  codes = get_codes(root)
  print(upper, '-->', get_encoded_size(codes))
  cr = round(report(codes))
  results[upper - 1] = cr # adjust since upper is the number of primes, -1 to make it just odd primes

primes = realprimes_up_to(10001)
results = {}
for n in range(3, 33):
    test_up_to(primes, n, results)

results

[5, 7, 11]
[2, 2, 4]
GapNode('/gn2', count=3, gap=0)
├── GapNode('/gn2/gn1', count=1, gap=4)
└── GapNode('/gn2/gn0', count=2, gap=2)
3 --> 3
[5, 7, 11, 13]
[2, 2, 4, 2]
GapNode('/gn5', count=4, gap=0)
├── GapNode('/gn5/gn4', count=1, gap=4)
└── GapNode('/gn5/gn3', count=3, gap=2)
4 --> 4
[5, 7, 11, 13, 17]
[2, 2, 4, 2, 4]
GapNode('/gn8', count=5, gap=0)
├── GapNode('/gn8/gn7', count=2, gap=4)
└── GapNode('/gn8/gn6', count=3, gap=2)
5 --> 5
[5, 7, 11, 13, 17, 19]
[2, 2, 4, 2, 4, 2]
GapNode('/gn11', count=6, gap=0)
├── GapNode('/gn11/gn10', count=2, gap=4)
└── GapNode('/gn11/gn9', count=4, gap=2)
6 --> 6
[5, 7, 11, 13, 17, 19, 23]
[2, 2, 4, 2, 4, 2, 4]
GapNode('/gn14', count=7, gap=0)
├── GapNode('/gn14/gn13', count=3, gap=4)
└── GapNode('/gn14/gn12', count=4, gap=2)
7 --> 7
[5, 7, 11, 13, 17, 19, 23, 29]
[2, 2, 4, 2, 4, 2, 4, 6]
GapNode('/gn19', count=8, gap=0)
├── GapNode('/gn19/gn15', count=4, gap=2)
└── GapNode('/gn19/gn18', count=4, gap=0)
    ├── GapNode('/gn19/gn18/gn17', count=1,

{2: 0,
 3: 0,
 4: 0,
 5: 0,
 6: 0,
 7: 25,
 8: 28,
 9: 25,
 10: 23,
 11: 25,
 12: 23,
 13: 21,
 14: 20,
 15: 22,
 16: 21,
 17: 19,
 18: 21,
 19: 20,
 20: 19,
 21: 18,
 22: 0,
 23: 0,
 24: 2,
 25: 2,
 26: 4,
 27: 4,
 28: 32,
 29: 33,
 30: 32,
 31: 32}

In [12]:
tmp = pyprimesieve.primes_nth(10000000)
primes = pyprimesieve.primes(tmp+1)
gaps = [*map(lambda i:primes[i]-primes[i-1], range(2,10000000))]

print(len(primes))

10000000


In [13]:
treeRoot = make_huffman_tree(gaps)
codes = get_codes(treeRoot)
codes

{2: ('1100', 738597),
 4: ('1101', 738717),
 6: ('100', 1297540),
 8: ('0101', 566151),
 10: ('1011', 729808),
 12: ('000', 920661),
 14: ('0100', 503524),
 16: ('11100', 371677),
 18: ('0111', 667734),
 20: ('10101', 354267),
 22: ('01101', 307230),
 24: ('0010', 453215),
 26: ('111110', 211203),
 28: ('111111', 229177),
 30: ('11101', 398713),
 32: ('001110', 123123),
 34: ('011000', 129043),
 36: ('111101', 206722),
 38: ('1010011', 94682),
 40: ('001100', 111546),
 42: ('101000', 159956),
 44: ('0011111', 64866),
 46: ('11110010', 54931),
 48: ('1010010', 93693),
 50: ('11110001', 52183),
 52: ('01100111', 38800),
 54: ('0011110', 64157),
 56: ('01100100', 32224),
 58: ('00110101', 27985),
 60: ('11110011', 55305),
 62: ('011001010', 16763),
 64: ('011001100', 17374),
 66: ('00110110', 30960),
 68: ('1111000010', 12368),
 70: ('011001101', 17475),
 72: ('011001011', 17255),
 74: ('0011011111', 8540),
 76: ('0011011101', 7253),
 78: ('001101000', 13758),
 80: ('11110000111', 6760),


In [14]:
fixed = get_fixed_size(codes)
encoded = get_encoded_size(codes)
print(fixed)
print(encoded)

69999986
44712373


# TODO Achieve Target Compression Ratios

In [24]:
print("From fixed:", report(codes))

From fixed: 36.12516865360516


In [31]:
def get_bin_size(codes):
  return sum([32 * count for gap, (code, count) in codes.items()])

print("From binary:", compression_ratio(get_bin_size(codes), get_encoded_size(codes)))

From binary: 86.02738064297613


In [56]:
def get_ascii_size(primes):
  return sum([len(str(prime) + '\n') * 8 for prime in primes])

print("From ASCII:", compression_ratio(get_ascii_size(primes), get_encoded_size(codes)))

From ASCII: 94.02141572742846


Your solution should correctly compute the following three compression ratios:

| Ratio       | Value              |
|-------------|--------------------|
| From fixed  | 36.125168653605158 |
| From binary |              86.03 |
| From ASCII  |              94.02 | 


# TODO My Report on What I Did and What I Learned

## Fun


I had fun on this assignment. I thought that this assignment was very difficult. However, I was able to meet with a group of students to work on the assignmnet, which made it more fun.

## New

In completing this assignment, I learned a lot more about huffman endoding. Before doing this assignment, I didn't know how to caculate a compression ratio. I know feel like I have a good understanding of compression ratios and how they are calculated. 

## Meaningful


I found this assignment meaningful. I think that I can build upon this information because I still do not have a lot of knowledge of Huffman Coding and Huffman Trees. I think that this knowledge will be valuable to me in setting up my understanding for future courses.

## Other

Connections: I can connect this knowledge to what I learned in DM1. In DM1 I did a little bit with Huffman encoding but have learned a lot more in this class!

Collaborators: Claire Hocker, Daniel Strickland, Betton Steiner. We met together in a zoom meeting and worked on the assignment together. 

# TODO What is True?
Click on each warranted checkbox to toggle it to True (or back to False). 

NOTE: *This only works in Colab. If you run it in some other Jupyter notebook client/server environment you may have to change False to True (or vice versa) manually.*

This self-assessment is subject to revision by a grader.

In [57]:
#@markdown ## What is True about what I did?
#@markdown ### I had fun.
cb00 = True #@param {type:'boolean'}
#@markdown ### I learned something new.
cb01 = True #@param {type:'boolean'}
#@markdown ### I achieved something meaningful, or something I can build upon at a later time.
cb02 = True #@param {type:'boolean'}
#@markdown ## What is True about my report?
#@markdown ### I wrote a sufficient number of well-written sentences.
cb03 = True #@param {type:'boolean'}
#@markdown ### My report is free of mechanical infelicities.
cb04 = True #@param {type:'boolean'}
#@markdown ### I used Grammarly (or something better described in my report) to check for MIs.
cb05 = True #@param {type:'boolean'}
#@markdown ### I reported on any connections I found between these problems and something I already know. 
cb06 = True #@param {type:'boolean'}
#@markdown ### I reported who were and what contribution each of my collaborators made.
cb07 = True #@param {type:'boolean'}
#@markdown ## What is True about my calculations?
#@markdown ### I correctly calculated the number of times each gap size occurs. 
cb08 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the number of bits per gap size with a fixed encoding.
cb09 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the total number of bits encoded with the Huffman encoding.
cb10 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the total number of bits encoded with the fixed encoding.
cb11 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the compression ratio from this fixed encoding.
cb12 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the size of the first ten million primes encoded as 32-bit integer binary data.
cb13 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the compression ratio from the binary size.
cb14 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the size of the first ten million primes encoded as ASCII data.
cb15 = True #@param {type:'boolean'}
#@markdown ### I correctly calculated the compression ratio from the ASCII size (just the primes, nothing else).
cb16 = True #@param {type:'boolean'}

## DO NOT CHANGE ANYTHING IN THE NEXT CODE CELL!!
### Delete this cell and the following ones before submitting your work.

In [58]:
points_for_what_I_did = [5]*3
points_for_my_report = [8]*5
points_for_my_calculations = [5]*9
points = points_for_what_I_did + points_for_my_report + points_for_my_calculations
# cb is short for checkbox
total = sum(map(lambda n, p: p if eval(f'cb{n:02}') else 0,
                range(len(points)), points))             
total

100

# For graders

In [ ]:
#@markdown ---
number_of_MIs_found = 0 #@param {type: 'slider', min: 0, max: 5}
#@markdown ---
